File: vinylkeeper_back/app/core/security.py
from fastapi.middleware.cors import CORSMiddleware
from fastapi import FastAPI
from app.core.config_env import settings
from argon2 import PasswordHasher
from argon2.exceptions import VerifyMismatchError
import logging

logger = logging.getLogger(__name__)

def configure_cors(app: FastAPI):
    origins = settings.ALLOWED_ORIGINS.split(",")
    app.add_middleware(
        CORSMiddleware,
        allow_origins=origins,
        allow_credentials=True,
        allow_methods=["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"],
        allow_headers=[
            "Content-Type",
            "Authorization",
            "Accept",
            "Origin",
            "X-Requested-With",
            "Access-Control-Request-Method",
            "Access-Control-Request-Headers"
        ],
        expose_headers=[
            "Content-Type",
            "Authorization"
        ],
        max_age=3600,
    )


ph = PasswordHasher()


def hash_password(password: str) -> str:
    return ph.hash(password)


def verify_password(hashed_password: str, plain_password: str) -> bool:
    try:
        ph.verify(hashed_password, plain_password)
        return True
    except VerifyMismatchError:
        logger.warning("Password verification failed: VerifyMismatchError")
        return False
    except Exception as e:
        logger.error(f"Password verification failed with error: {str(e)}")
        logger.error(f"Error type: {type(e).__name__}")
        logger.error(f"Error details: {e}")
        return False

File: vinylkeeper_back/src/core/security.rs
use rocket::http::Method;
use rocket_cors::{AllowedHeaders, AllowedOrigins, Cors, CorsOptions};
use std::env;

pub fn create_cors_fairing() -> Cors {
    let allowed_origins_env = env::var("ALLOWED_ORIGINS").unwrap_or_else(|_| String::from(""));
    let allowed_origins: Vec<&str> = allowed_origins_env.split(',').collect();

    let allowed_origins = AllowedOrigins::some_exact(&allowed_origins);

    CorsOptions {
        allowed_headers: AllowedHeaders::some(&["Content-Type", "Authorization", "Accept"]),
        allowed_origins,
        allowed_methods: vec![
            Method::Get,
            Method::Post,
            Method::Put,
            Method::Patch,
            Method::Delete,
            Method::Options,
            Method::Head,
        ]
        .into_iter()
        .map(Into::into)
        .collect(),
        allow_credentials: true,
        expose_headers: vec!["Content-Type"].into_iter().map(Into::into).collect(),
        max_age: Some(3600),
        ..Default::default()
    }
    .to_cors()
    .expect("Failed to create CORS fairing")
}
